components/tutorial/tutorial_tip/tutorial_tip.tsx

```diff
@@ -159,15 +173,39 @@ export default class TutorialTip extends React.PureComponent<Props, State> {
         }
     }
 
-    private handleSavePreferences = (autoTour: boolean, nextStep: boolean | number): void => {
-        const {currentUserId, tutorialCategory, actions, singleTip, onNextNavigateTo, onPrevNavigateTo} = this.props;
+    getKeyByValue = (obj: Record<string, number>, value: number) => {
+        return Object.keys(obj).find((key) => obj[key] === value);
+    }
+
+    handleSavePreferences = (autoTour: boolean, nextStep: boolean | number): void => {
+        const {isAdmin, currentUserId, tutorialCategory, actions, singleTip, onNextNavigateTo, onPrevNavigateTo} = this.props;
         const {closeRhsMenu, savePreferences, setFirstChannelName} = actions;
 
         let stepValue = this.props.currentStep;
         if (nextStep === true) {
             stepValue += 1;
+
+            // if the next tip step/steps are for only admins, skip them for non admins
+            for (const tipName of AdminOnBoardingTutorialStep) {
+                const keyForNextStepValue = this.getKeyByValue(OnBoardingTutorialStep, stepValue);
+                if (!isAdmin && keyForNextStepValue && keyForNextStepValue === tipName) {
+                    stepValue += 1;
+                } else {
+                    break;
+                }
+            }
         } else if (nextStep === false) {
             stepValue -= 1;
+
+            // if the previous tip step/steps are for only admins, skip them for non admins
+            for (const tipName of AdminOnBoardingTutorialStep) {
```

Comment:
Nice solution to skipping admin steps!! For this backing up steps code path, I think we need to iterate in reverse order, which we could do with

```ts
for (const tipName of [...AdminOnBoardingTutorialStep].reverse()) {
...
```
My code example creates a new array because reverse mutates the original array, which we do not want to do.

Reply:
@neallred why do we need to reverse the array?

Reply:
If there are two steps in a row that are admin steps and you are backing up and iterating from start to finish, when the first admin step is reached, in the for loop iteration, the user will still be on the second admin step. Then, once the loop reaches the second admin step, the step will be backed up once. But even though the user is still on an admin step, iteration over the step the user is on is already finished, so the user is incorrectly placed on an admin step.

That's kind of a lot to step through, so I made a minimal code repro where you can see it happening.

```js
const TutorialSteps = {
    ADD_FIRST_CHANNEL: -1,
    POST_POPOVER: 0,
    CHANNEL_POPOVER: 1,
    ADD_CHANNEL_POPOVER: 2,
    MENU_POPOVER: 3,
    PRODUCT_SWITCHER: 4,
    SETTINGS: 5,
    START_TRIAL: 6,
    NEW_ADMIN_STEP: 7,
    OTHER_NEW_STEP: 8,
    FINISHED: 999,
};

const ADMIN_STEPS = ["START_TRIAL", "NEW_ADMIN_STEP"]

function getKeyByValue(obj, value) {
    return Object.keys(obj).find((key) => obj[key] === value);
}

function backupStepsNoReverse(currentStep, isAdmin) {
    let newStep = currentStep - 1;
    if (isAdmin) {
        return newStep;
    }
    for (const adminStep of ADMIN_STEPS) {
        if (getKeyByValue(TutorialSteps, newStep) === adminStep) {
            newStep -= 1;
        }
    }
    return newStep
}

function backupSteps(currentStep, isAdmin) {
    let newStep = currentStep - 1;
    if (isAdmin) {
        return newStep;
    }
    for (const adminStep of [...ADMIN_STEPS].reverse()) {
        if (getKeyByValue(TutorialSteps, newStep) === adminStep) {
            newStep -= 1;
        }
    }
    return newStep
}

console.log(backupStepsNoReverse(TutorialSteps.OTHER_NEW_STEP, false));
console.log(backupSteps(TutorialSteps.OTHER_NEW_STEP, false));
```

If you run that as a file in node, you get:
```
node tmp.js
6
5
```

Reply:
Hmmm... no problem changing it but currently I have a test like so, if we are currently on step 8, but step 7 and 6 are admin steps, then we expect to land on step 5

tutorial_tip.test.tsx
```js
    test('handleSavePreferences skip previous admin steps for non admins', () => {
        const props = {
            currentUserId,
            title: <>{'title'}</>,
            step: 8,
            currentStep: 8,
            autoTour: false,
            isAdmin: false,
            actions: {
                closeRhsMenu: jest.fn(),
                savePreferences: jest.fn(),
                setFirstChannelName: jest.fn(),
                setProductMenuSwitcherOpen: jest.fn(),
            },
            screen: <></>,
            placement: 'right',

        };
        const wrapper = shallow<TutorialTip>(
            <TutorialTip {...props}/>,
        );

        wrapper.instance().handleSavePreferences(false, false);

        // current tip is 7, but tip 6, START_TRIAL is only for admins and is skipped to tip 5
        expect(props.actions.savePreferences).toHaveBeenCalledWith('currentUserId', [{category: 'tutorial_step', name: 'currentUserId', user_id: 'currentUserId', value: '5'}]);
    });
```

and with constants like so
```js
export const TutorialSteps = {
    ADD_FIRST_CHANNEL: -1,
    POST_POPOVER: 0,
    CHANNEL_POPOVER: 1,
    ADD_CHANNEL_POPOVER: 2,
    MENU_POPOVER: 3,
    PRODUCT_SWITCHER: 4,
    SETTINGS: 5,
    START_TRIAL: 6,
    NEW_ADMIN_STEP: 7,
    FINISHED: 999,
};

export const AdminTutorialSteps = ['NEW_ADMIN_STEP', 'START_TRIAL'];
```

Reply:
nvm, I actually see what you mean

Reply:
I should probably document that the order the tips are added to `AdminTutorialSteps` matters